subprojects/enterprise/src/main/java/org/gradle/internal/tasks/impl/DefaultTaskProperties.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.tasks.impl;
+
+import org.gradle.internal.tasks.InputFileProperty;
+import org.gradle.internal.tasks.OutputFileProperty;
+import org.gradle.internal.tasks.TaskProperties;
+
+import java.util.Set;
+
+public class DefaultTaskProperties implements TaskProperties {
+
+    private final Set<InputFileProperty> inputFiles;
+    private final Set<OutputFileProperty> outputFiles;
+
+    DefaultTaskProperties(Set<InputFileProperty> inputFiles, Set<OutputFileProperty> outputFiles) {
+        this.inputFiles = inputFiles;
+        this.outputFiles = outputFiles;
+    }
```

Comment:
In general storing references to potentially mutable collections sounds like a bad idea. Do we have some other mechanism to prevent the related issues?

Reply:
This is all preliminary but you're right we should make sure they're immutable here.

Reply:
Done in da8e1d8f0adacfd652359f9bdac7804827d0d2c6.